sdks/python/apache_beam/io/mongodbio.py

```diff
@@ -66,542 +66,771 @@
+            ]
+            buckets = list(
+                # Use `allowDiskUse` option to avoid aggregation limit of 100 Mb RAM
+                client[self.db][self.coll].aggregate(pipeline, allowDiskUse=True)
+            )
+            if buckets:
+                buckets[-1]["_id"]["max"] = end_pos
+
+            return buckets
+
+    def _merge_id_filter(
+        self,
+        start_position: Union[int, str, ObjectId],
+        stop_position: Union[int, str, ObjectId] = None,
+    ) -> dict:
+        """Merge the default filter (if any) with refined _id field range
+        of range_tracker.
+        $gte specifies start position (inclusive)
+        and $lt specifies the end position (exclusive),
+        see more at
+        https://docs.mongodb.com/manual/reference/operator/query/gte/ and
+        https://docs.mongodb.com/manual/reference/operator/query/lt/
+        """
+        if stop_position is None:
+            id_filter = {"_id": {"$gte": start_position}}
+        else:
+            id_filter = {"_id": {"$gte": start_position, "$lt": stop_position}}
+
+        if self.filter:
+            all_filters = {
+                # see more at
+                # https://docs.mongodb.com/manual/reference/operator/query/and/
+                "$and": [self.filter.copy(), id_filter]
+            }
+        else:
+            all_filters = id_filter
+
+        return all_filters
+
+    def _get_head_document_id(self, sort_order):
+        with MongoClient(self.uri, **self.spec) as client:
+            cursor = (
+                client[self.db][self.coll]
+                .find(filter={}, projection=[])
+                .sort([("_id", sort_order)])
+                .limit(1)
+            )
+            try:
+                return cursor[0]["_id"]
+
+            except IndexError:
+                raise ValueError("Empty Mongodb collection")
+
+    def _replace_none_positions(self, start_position, stop_position):
+
+        if start_position is None:
+            start_position = self._get_head_document_id(ASCENDING)
+        if stop_position is None:
+            last_doc_id = self._get_head_document_id(DESCENDING)
+            # increment last doc id binary value by 1 to make sure the last document
+            # is not excluded
+            stop_position = _ObjectIdHelper.increment_id(last_doc_id, 1)
+
+        return start_position, stop_position
+
+    def _count_id_range(self, start_position, stop_position):
+        """Number of documents between start_position (inclusive)
+        and stop_position (exclusive), respecting the custom filter if any.
+        """
+        with MongoClient(self.uri, **self.spec) as client:
+            return client[self.db][self.coll].count_documents(
+                filter=self._merge_id_filter(start_position, stop_position)
+            )
+
+
+class MongoDBRangeTracker(OrderedPositionRangeTracker):
+    """RangeTracker for tracking MongoDB `_id` of following types:
+    - int
+    - bytes
```

Comment:
can you add unit test and integration test accordingly to make sure they work as intended?

Reply:
defeinitely! I will

Reply:
unittests added